rclcpp/multithreaded_executor/multithreaded_executor.cpp

```diff
@@ -0,0 +1,177 @@
+ * We will have a single subscriber node running 2 threads. Each thread loops at different speeds, and
+ * just repeats what it sees from the publisher to the screen.
+ */
+
+class PublisherNode : public rclcpp::Node
+{
+public:
+  PublisherNode()
+  : Node("PublisherNode")
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello World! " + std::to_string(this->count_++);
+
+        // Extract current thread
+        auto curr_thread = string_thread_id();
+
+        // Prep display message
+        auto info_message = "\n<<THREAD " + curr_thread + ">> Publishing '%s'";
+        RCLCPP_INFO(this->get_logger(), info_message, message.data.c_str());
+        this->publisher_->publish(message);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+class DualThreadedNode : public rclcpp::Node
+{
+public:
+  DualThreadedNode()
+  : Node("DualThreadedNode")
+  {
+    /* These define the callback groups
+     * They don't really do much on their own, but they have to exist in order to
+     * assign callbacks to them. They're also what the executor looks for when trying to run multiple threads
+     */
+    callback_group_subscriber1_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+    callback_group_subscriber2_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+
+    // Each of these callback groups is basically a thread
+    // Everything assigned to one of them gets bundled into the same thread
+    auto sub1_opt = rclcpp::SubscriptionOptions();
+    sub1_opt.callback_group = callback_group_subscriber1_;
+    auto sub2_opt = rclcpp::SubscriptionOptions();
+    sub2_opt.callback_group = callback_group_subscriber2_;
+
+    subscription1_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10),
+      // std::bind is sort of C++'s way of passing a function
+      // If you're used to function-passing, skip these comments
+      std::bind(
+        &DualThreadedNode::subscriber1_cb,  // First parameter is a reference to the function
+        this,                               // What the function should be bound to
+        std::placeholders::_1),             // At this point we're not positive of all the
+                                            // parameters being passed
+                                            // So we just put a generic placeholder
+                                            // into the binder
+                                            // (since we know we need ONE parameter)
+      sub1_opt);                  // This is where we set the callback group.
```

Comment:
this argument trailing so far down here is a bit odd to me, and it also looks like the indentation is off.

Could you add a block comment above it to explain?

Or, frankly, I'm not so sure it's worth explaining `std::bind` in this example, since it is not exclusively in this example. In other words, this comment, if deemed necessary, should probably go in a more simple example, since the knowledge of what `std::bind` does should be pretty clear by this point when looking through examples.

Reply:
I could move the side comments to a block comment above the call, sure. I felt that having them on the side of what they're commenting about was a bit more legible. Still, give me a few minutes, and I'll have that out.

I feel that explaining the `std::bind` is worth it here though. ROS 2 is not supposed to be used just by people who are extremely knowledgeable about the deeper concepts of C++ programming. Many will be Electrical Engineers, Mechanical Engineers, or undergraduates, all with varying levels of programming expertise. As such I have no way of guaranteeing that they all will understand what `std::bind` does. I felt that adding clarifying comments would minimize headaches all around.

Reply:
It's a little verbose, but I think it is worthwhile to have an explanation of `std::bind` here alongside how we use it in ROS 2.  So I'm fine with keeping it.

Reply:
> It's a little verbose, but I think it is worthwhile to have an explanation of std::bind here alongside how we use it in ROS 2

Yeah, makes sense to me.